File: solutions/r/protein-translation/protein-translation.R
translate <- function(bases) {
  if (nchar(bases) == 0) {
    return(c())
  }

  group_chunks <- function(strand) {
    # group_chunks returns a vector of strings with max length 3
    regmatches(strand, gregexpr(".{1,3}", strand)) |> unlist()
  }

  from_codon <- function(codon) {
    switch(codon,
      AUG = "Methionine",
      UUU = ,
      UUC = "Phenylalanine",
      UUA = ,
      UUG = "Leucine",
      UCU = ,
      UCC = ,
      UCA = ,
      UCG = "Serine",
      UAU = ,
      UAC = "Tyrosine",
      UGU = ,
      UGC = "Cysteine",
      UGG = "Tryptophan",
      UAA = ,
      UAG = ,
      UGA = "STOP"
    )
  }

  result <- c()
  for (chunk in group_chunks(bases)) {
    amino_acid <- from_codon(chunk)
    if (amino_acid == "STOP") break
    result <- c(result, amino_acid)
  }
  result
}

File: solutions/nim/protein-translation/protein_translation.nim
proc translate*(s: string): seq[string] =
  for i in countup(0, s.high, 3):
    case s[i..<min(i+3, s.len)]:
      of "AUG": result.add("Methionine")
      of "UUU", "UUC": result.add("Phenylalanine")
      of "UUA", "UUG": result.add("Leucine")
      of "UCU", "UCC", "UCA", "UCG": result.add("Serine")
      of "UAU", "UAC": result.add("Tyrosine")
      of "UGU", "UGC": result.add("Cysteine")
      of "UGG": result.add("Tryptophan")
      of "UAA", "UAG", "UGA": break
      else: raise newException(ValueError, "invalid codon")
